White-paper (#438)
cqfn/diKTat: wp/sections/compare.tex

\subsection{About Ktlint}

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle.

Reviewer: tell more that it is a framework. It can help you to build a custom set of rules

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle.

Reviewer: write that ktlint code style has strictly hardcoded rules

\subsection{About Detekt}

> \par Detekt is a static code analysis tool. It operates on an abstract syntax tree provided by Kotlin compiler. Detekt supports such features as code smell analysis, highly configurable rule sets, IntelliJ integration and third-party integrations for Maven, Bazel and Github actions and many more. It has been developing since 2016 and today it has 3.2k stars, 411 forks and 1850 closed PRs. It has around 45k lines of code.

Reviewer: as I was told in ktlint - detekt is using it for some of rules. 
Could you please check it and write it here?

Reviewer 2: Yes, detekt has a whole group of rules, `detekt-formatting-rules`, that are wrappers around ktlint rules

> \subsection{About diKTat}

Reviewer: how about talking about Intellij Idea. Say something like that:
"Jetbrains created  Kotlin and Idea, so they added static analysis to IDE. But this is not well configurable and is not useful for CI/CD process."

> \subsection{About diKTat}

Reviewer: We forget that diktat is very needed as a CI/CD tool, because no one looks at IDE warnings and we need a tool that will prevent merging bullshit into the code

> \par So why are we better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Secondly, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Finally, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out.

Reviewer: we - change to diktat

> \par So why are we better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Secondly, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Finally, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out.

Reviewer: first of all describe diktat. After that say "so why it is better?"

> \par So why are we better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Secondly, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Finally, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out.

Reviewer: write phrases about finding "code smells", "code style issues", "bugs that can be founf by complex AST analysis"

> Ideas: may be find ktlint and detekt in big projects, use insights to draw graphs. 

Reviewer: good idea! ktlint has the list of users in their code. Not in readme, but somewhere in users/contributors, something like this

\subsection{About ktlint}

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle. To configure rules in Ktlint you should modify .editorconfig file. But you can't configure specific rules, instead you can provide some common settings. Properties should be specified under \textbf{[*.{kt,kts}]}. If you want to implement your own rules you need to create a ruleset. Ktlint is using java's ServiceLoader to discover all available "RuleSets". Basically you should provide implementation of "RuleSetProvider" interface. Ktlint refers to article on \textbf{medium} \footnote{\url{https://medium.com/@vanniktech/writing-your-first-ktlint-rule-5a1707f4ca5b}}on how to create a ruleset and a rule.

Reviewer: created by pinterest

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle. To configure rules in Ktlint you should modify .editorconfig file. But you can't configure specific rules, instead you can provide some common settings. Properties should be specified under \textbf{[*.{kt,kts}]}. If you want to implement your own rules you need to create a ruleset. Ktlint is using java's ServiceLoader to discover all available "RuleSets". Basically you should provide implementation of "RuleSetProvider" interface. Ktlint refers to article on \textbf{medium} \footnote{\url{https://medium.com/@vanniktech/writing-your-first-ktlint-rule-5a1707f4ca5b}}on how to create a ruleset and a rule.

Reviewer: write more about Service Loader, something like this: "ServiceLoader is used to inject your own implementation of rules for the static analysis. In this case ktlint becomes a third-party dependency and a framework."

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle. To configure rules in Ktlint you should modify .editorconfig file. But you can't configure specific rules, instead you can provide some common settings. Properties should be specified under \textbf{[*.{kt,kts}]}. If you want to implement your own rules you need to create a ruleset. Ktlint is using java's ServiceLoader to discover all available "RuleSets". Basically you should provide implementation of "RuleSetProvider" interface. Ktlint refers to article on \textbf{medium} \footnote{\url{https://medium.com/@vanniktech/writing-your-first-ktlint-rule-5a1707f4ca5b}}on how to create a ruleset and a rule.

Reviewer: It checks and can automatically fix the code.

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle. To configure rules in Ktlint you should modify .editorconfig file. But you can't configure specific rules, instead you can provide some common settings. Properties should be specified under \textbf{[*.{kt,kts}]}. If you want to implement your own rules you need to create a ruleset. Ktlint is using java's ServiceLoader to discover all available "RuleSets". Basically you should provide implementation of "RuleSetProvider" interface. Ktlint refers to article on \textbf{medium} \footnote{\url{https://medium.com/@vanniktech/writing-your-first-ktlint-rule-5a1707f4ca5b}}on how to create a ruleset and a rule.

Reviewer: write that they have a "fixed hardcoded" codestyle

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle. To configure rules in Ktlint you should modify .editorconfig file. But you can't configure specific rules, instead you can provide some common settings. Properties should be specified under \textbf{[*.{kt,kts}]}. If you want to implement your own rules you need to create a ruleset. Ktlint is using java's ServiceLoader to discover all available "RuleSets". Basically you should provide implementation of "RuleSetProvider" interface. Ktlint refers to article on \textbf{medium} \footnote{\url{https://medium.com/@vanniktech/writing-your-first-ktlint-rule-5a1707f4ca5b}}on how to create a ruleset and a rule.

Reviewer: write  how it can be run: maven/gradle/cli

> \par Ktlint is a popular an anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint has been developing since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which divides on standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle. To configure rules in Ktlint you should modify .editorconfig file. But you can't configure specific rules, instead you can provide some common settings. Properties should be specified under \textbf{[*.{kt,kts}]}. If you want to implement your own rules you need to create a ruleset. Ktlint is using java's ServiceLoader to discover all available "RuleSets". Basically you should provide implementation of "RuleSetProvider" interface. Ktlint refers to article on \textbf{medium} \footnote{\url{https://medium.com/@vanniktech/writing-your-first-ktlint-rule-5a1707f4ca5b}}on how to create a ruleset and a rule.

Reviewer 2: ```suggestion
\par Ktlint is a popular anti-bikeshedding Kotlin linter with built-in formatter. It tries to capture (reflect) official code style from kotlinlang.org and Android Kotlin Style Guide and then automatically apply these rules to your codebase. It claims to be simple and easy to use. Ktlint is being developed since 2016 and from then on it has 3.8k stars, 299 forks and 390 closed PRs (at least on the moment of writing this whitepaper). There have been written over 15k lines of code. Ktlint has it's own ruleset, which is divided into standard and experimental rules. Ktlint can be used as a plugin via Maven or Gradle. To configure rules in Ktlint you should modify .editorconfig file. But you can't configure specific rules, instead you can provide some common settings. Properties should be specified under \textbf{[*.{kt,kts}]}. If you want to implement your own rules you need to create a ruleset. Ktlint is using java's ServiceLoader to discover all available "RuleSets". Basically you should provide implementation of "RuleSetProvider" interface. Ktlint refers to article on \textbf{medium} \footnote{\url{https://medium.com/@vanniktech/writing-your-first-ktlint-rule-5a1707f4ca5b}}on how to create a ruleset and a rule.
```
maybe specific date instead of 'moment of writing'?
also, IIRC ktlint's plugins are not official, maybe it's worth mentioning?

> \par A lot of projects uses ktlint as their code formatting tool. For example, OmiseGo \footnote{\url{https://github.com/omgnetwork/android-sdk}} (currently rebranding to OMG Network) - a quite popular cryptocurrency.

Reviewer: This is an alphabetical list of people and organizations who are using this
project. If you'd like to be included here, please send a Pull Request that
adds your information to this file.

- [Airbnb](https://www.airbnb.com/)
- [Ellation](http://www.ellation.com/)
- [Faithlife](https://faithlife.com/about)
- [Freeletics](https://www.freeletics.com)
- [Gradle](https://gradle.org)
- [Pinterest](https://www.pinterest.com/)
- [Shopify](https://www.shopify.com/)
- [Tinder](https://tinder.com)
- [Workable](https://www.workable.com)

\subsection{About detekt}

> \par Detekt is a static code analysis tool. It operates on an abstract syntax tree provided by Kotlin compiler. As it was mentioned in ktlint section, it has it's own rules. Detekt uses ktlint to redefine it's formatting rules. Detekt supports such features as code smell analysis, highly configurable rule sets, IntelliJ integration and third-party integrations for Maven, Bazel and Github actions and many more. It has been developing since 2016 and today it has 3.2k stars, 411 forks and 1850 closed PRs. It has around 45k lines of code.

Reviewer: not only AST. They do more complex analysis

> \par Detekt is a static code analysis tool. It operates on an abstract syntax tree provided by Kotlin compiler. As it was mentioned in ktlint section, it has it's own rules. Detekt uses ktlint to redefine it's formatting rules. Detekt supports such features as code smell analysis, highly configurable rule sets, IntelliJ integration and third-party integrations for Maven, Bazel and Github actions and many more. It has been developing since 2016 and today it has 3.2k stars, 411 forks and 1850 closed PRs. It has around 45k lines of code.

Reviewer: has a mechanism for suppression of warnings

> \par Detekt is a static code analysis tool. It operates on an abstract syntax tree provided by Kotlin compiler. As it was mentioned in ktlint section, it has it's own rules. Detekt uses ktlint to redefine it's formatting rules. Detekt supports such features as code smell analysis, highly configurable rule sets, IntelliJ integration and third-party integrations for Maven, Bazel and Github actions and many more. It has been developing since 2016 and today it has 3.2k stars, 411 forks and 1850 closed PRs. It has around 45k lines of code.

Reviewer 2: ```suggestion
\par Detekt is a static code analysis tool. It operates on an abstract syntax tree provided by Kotlin compiler. Similarly to ktlint, it has it's own rules. Detekt uses wrapped ktlint rules as it's formatting rules. Detekt supports such features as code smell analysis, highly configurable rule sets, IntelliJ integration and third-party integrations for Maven, Bazel and Github actions and many more. It is being developed since 2016 and today it has 3.2k stars, 411 forks and 1850 closed PRs. It has circa 45k lines of code.
```

> \par Detekt is used in such projects as fountain or Kaspresso. Fountain "is an Android Kotlin library conceived to make your life easier when dealing with paged endpoint services" \footnote{\url{https://github.com/xmartlabs/fountain}} and Kaspresso is a framework for UI testing on Android made by KasperskyLab \footnote{\url{https://github.com/KasperskyLab/Kaspresso}}.

Reviewer: names like "fountain" should be in quotes ("")

\subsection{About ktfmt}

> \par Ktfmt is a program that formats Kotlin code, based on google-java-format. It's development has started in Facebook in the end of 2019. It can be added to your project through a Maven dependency, Gradle dependency, IntelliJ plugin or you can run it through a command line. Ktfmt is not a configurable, so to change any rule logic you need to download the project and redefine some constants. Ktfmt has 214 stars, 16 forks, 20 closed PRs and around 7500 lines of code. 

Reviewer: created by facebook

> \par Ktfmt is a program that formats Kotlin code, based on google-java-format. It's development has started in Facebook in the end of 2019. It can be added to your project through a Maven dependency, Gradle dependency, IntelliJ plugin or you can run it through a command line. Ktfmt is not a configurable, so to change any rule logic you need to download the project and redefine some constants. Ktfmt has 214 stars, 16 forks, 20 closed PRs and around 7500 lines of code. 

Reviewer: "is not a configurable **application**"

\subsection{About diKTat}

> \par Diktat as well as ktlint and detekt is a static code analysis tool. It's development started in 2020 and at the time of writing this whitepaper diKTat has 130 stars and 12 forks. DiKTat operates on AST provided by IntelliJ. 

Reviewer: diktat not only a tool, but also coding convention (link) that in details describes all the rules that you should follow when writing a code on Kotlin

> \par Diktat as well as ktlint and detekt is a static code analysis tool. It's development started in 2020 and at the time of writing this whitepaper diKTat has 130 stars and 12 forks. DiKTat operates on AST provided by IntelliJ. 

Reviewer 2: ```suggestion
\par Diktat as well as ktlint and detekt is a static code analysis tool. It's development started in 2020 and at the time of writing this whitepaper diKTat has 130 stars and 12 forks. DiKTat operates on AST provided by kotlin compiler. 
```
You say like this about AST for detekt, so I think we should be consistent.

> \par So why is diKTat better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Second, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Third, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out. Last but not the least, diKTat can be used as a CI/CD tool in order to avoid merging errors in the code.

Reviewer: 100 complex rules: that can both check and fix your code

> \par So why is diKTat better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Second, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Third, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out. Last but not the least, diKTat can be used as a CI/CD tool in order to avoid merging errors in the code.

Reviewer: "is commented out with the description"

> \par So why is diKTat better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Second, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Third, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out. Last but not the least, diKTat can be used as a CI/CD tool in order to avoid merging errors in the code.

Reviewer: >>> you can choose whether you need a copyright or choose a length of line
you have taken the most primitive rules?))

> \par So why is diKTat better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Second, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Third, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out. Last but not the least, diKTat can be used as a CI/CD tool in order to avoid merging errors in the code.

Reviewer 2: How many rules does ktlint have? We should specify it too if we are using our figure.

Reviewer 3: Number of rules are specified in the table. Should we mention their number in each subsection?

Reviewer 2: I guess I didn't see the table, where is it? I just thought that if we say that we have 100 rules which is more than ktlint has, we should say exactly how many does ktlint have.

> \par So why is diKTat better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Second, diKTat is configurable. A lot of rules have it's own settings, which can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Third, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out. Last but not the least, diKTat can be used as a CI/CD tool in order to avoid merging errors in the code.

Reviewer 2: ```suggestion
\par So why is diKTat better? First of all, we support much more rules than ktlint. Our ruleset includes more than 100 rules. Second, diKTat is configurable. A lot of rules have their own settings, and all of them can be easily understood. For example, you can choose whether you need a copyright or choose a length of line. Third, diKTat is very easy to configure. You don't need to spend hours only to understand what each rule is doing. Our ruleset is a yml file, where each rule is commented out. Last but not the least, diKTat can be used as a CI/CD tool in order to avoid merging errors in the code.
```

\par Overall it can find code smells and code style issues. Also it can find pretty unobvious bugs by complex AST analysis.

\subsection{A few words about Jetbrains}

> \par Jetbrains is an international company located in Prague, Czech Republic. It created one of the best IDEs for Java and Kotlin called IntelliJ. This IDE supports a built-in linter. However it is not a well-configurable tool and it is not useful for CI/CD.

Reviewer: > Jetbrains is an international company located in Prague, Czech Republic.
remove it :)

> \par Jetbrains is an international company located in Prague, Czech Republic. It created one of the best IDEs for Java and Kotlin called IntelliJ. This IDE supports a built-in linter. However it is not a well-configurable tool and it is not useful for CI/CD.

Reviewer: > However it is not a well-configurable tool and it is not useful for CI/CD.
However it is not a well-configurable tool, you are not able to specify your own coding convention and it is not useful for CI/CD as it is highly coupled with it's IDE.
